src/main/java/io/vertx/core/http/impl/ClientConnection.java

```diff
@@ -242,8 +248,12 @@ private void handshakeComplete(ChannelHandlerContext ctx, FullHttpResponse respo
       }
       ws = new WebSocketImpl(vertx, ClientConnection.this, supportsContinuation, client.getOptions().getMaxWebsocketFrameSize());
       handshaker.finishHandshake(channel, response);
+
       log.debug("WebSocket handshake complete");
-      wsConnect.handle(ws);
+      context.executeFromIO(() -> {
```

Comment:
handshake complete is executed from channelread which is already executed on the right thread so this appears to be doing it twice.